### ISSUE TITLE
AsyncStorage.getItem() doesn't seem to work

### ISSUE THREAD

**Issue by mrded** [facebook/react-native#18372]
When I try to set a value via `AsyncStorage.getItem()`, I cannot request it back.

### Environment

Environment:
  OS: macOS High Sierra 10.13.3
  Node: 9.8.0
  Yarn: 1.5.1
  npm: 5.6.0
  Watchman: 4.9.0
  Xcode: Xcode 9.2 Build version 9C40b
  Android Studio: Not Found

Packages: (wanted => installed)
  react: ^16.3.0-alpha.1 => 16.3.0-alpha.1
  react-native: 0.54.0 => 0.54.0

### Expected Behavior

    await AsyncStorage.setItem('foo', 'bar');
    await AsyncStorage.getItem('foo').then(console.log); // 'bar'
    await AsyncStorage.getAllKeys().then(console.log); // ['foo']

### Actual Behavior

    await AsyncStorage.setItem('foo', 'bar');
    await AsyncStorage.getItem('foo').then(console.log); // null
    await AsyncStorage.getAllKeys().then(console.log); // []

### Steps to Reproduce

    import { AsyncStorage } from 'react-native';

    it('should be able to request a value after it was set', async () => {
      await AsyncStorage.setItem('foo', 'bar');
      const output = await AsyncStorage.getItem('foo');

      expect(output).toBe('bar');
    });

    it('should be able to see a new key after a value was set', async () => {
      await AsyncStorage.setItem('foo', 'bar');
      const output = await AsyncStorage.getAllKeys();

      expect(output).toContain('foo');
    });

**Comment by mrded** [comment]
Updated `react-native` => 0.54.0

**Comment by hjJunior** [comment]
I'm getting the same issue, and I'm using 0.54.0 react-native version, so what I can do to solve this? is It the stable version ?

**Comment by xiaonull** [comment]
Try to run the install command again, it may to work.

**Comment by hjJunior** [comment]
@xiaonull did you mean I run `npm install` before run `react-native run-android`
If is it, I have tried, even I tried to unistall and after install aplication on the device

**Comment by arjasco** [comment]
Having the same issue with `0.52.0`. Really frustrating. 

If you call `setItem` and then immediately call `getItem` or `getAllKeys` to reveal all in the store it works as normal and the value will be present. Reload of the app and back to square one, getItem returns null.

**Comment by Metroninja** [comment]
Running into this issue sporadically on 0.54.2

**Comment by yashojha19** [comment]
It is really frustrating. after few time of running the app. Aysnc AsyncStorage.getItem() stops working.
I tried many hacky way to work around it. but after few times. its just don't work.

**Comment by arjasco** [comment]
RN is pretty new to me, anyone know any of the maintainers on here we can ping to have a look?

The implementation is definitely not stable, there is another issue here: https://github.com/facebook/react-native/issues/12830

Don't really fancy using SQLite just to store a couple of integer values 😕

**Comment by yashojha19** [comment]
I am thinking to use this as alternative : https://github.com/oblador/react-native-keychain

**Comment by lenoxzzedwin** [comment]
@AleskiWeb  that is very stressful.
It's happening to me.

Some alternative?

**Comment by almorak** [comment]
same here ...
let result = await AsyncStorage.getItem(key)
android just stuck there , try catch not work, only restart the app again or setting a timeout,
but it is not a solution. It still cannot load the data saved.

**Comment by nahidmbstu** [comment]
let value = await AsyncStorage.getItem("user");     
this line blocks the execution .... 
the next line is not working .....   
i am using  ..........  "react-native": "0.55.3"  and android one device....
its frustating........

**Comment by HosseinAlipour** [comment]
the same issue, is it supposed to work with expo app or the issue is global?

**Comment by ImadeLak** [comment]
Hello 
the same here..... 

"dependencies": {
    "expo": "^27.0.1",
    "react": "16.3.1",
    "react-native": "~0.55.2",
    "react-native-router-flux": "^4.0.0-beta.31"
  }

I don't understand why it's not working...

**Comment by Jalalx** [comment]
Any update on this?

**Comment by C0dekid** [comment]
Same issue, still waiting for a fix!

**Comment by demarchisd** [comment]
Same issue here.

**Comment by finnfiddle** [comment]
I had the same problem and I found the reason was I was just going to fast for what the bundler could handle. I was often saving several changes in several files in close succession and what @AleskiWeb described was happening. I had to slow down and wait for the bundler to finish rebundling each time I save a file before I could save another. Hope this helps someone

**Comment by kkusanagi** [comment]
Thanks for the hint.
Seems like we need to close the Hot Reloading and close the app. Then open back and it work.

**Comment by yashojha19** [comment]
What about redux-persist, is it good ?

**Comment by AleskiWeb** [comment]
Just as an aside, you don't actually have to close your emulator, or even your bundler/builder. Just long press the application (or go to the settings as per usual) open the settings and then "FORCE STOP" the application. If you then just re-open the application on the emulator, the localstorage is back and functioning again. Its been my sort-of-workaround for a while now.

**Comment by apcjs** [comment]
You can write a batch file to do the following

adb shell am force-stop com.<YOUR PACKAGE NAME>
adb shell am start com.<YOUR PACKAGE NAME>/com.<YOUR PACKAGE NAME>.MainActivity

Run this batch file instead of pressing RR on the emulator

**Comment by ilyabozhkov** [comment]
@yashojha19, redux-persist is pretty bad. Too much code for nothing. There is "redux-offline" being a superstructure, more like opinionated interface to popular but unwieldy redux-persist. It doesn't make it better, just because redux-persist is bad at the core.

**Comment by stonecold123** [comment]
````
try {
     console.log('start')
     await AsyncStorage.getItem('token');
     alert('end') // << This never executes!
}catch(err){
     console.log(err);
 }
````
Same problem. The `AsyncStorage.getItem()` and the code followed never execute. No error at all. 
I'm using react-native-navigation v2 and react native 0.56.

**Comment by kkusanagi** [comment]
@stonecold123 I'm not sure how script work. Just to confirm.

Did you add AsyncStorage class while import from React Native?

In order to use "await", did you add async function?
Exp: 

```
const fund = async (a)=>{

}; 

```

Did you close hot reloading? try close the app and start over again.

Remove the app and try build over again.

Should be work in some cases. For me, I just close app and open again.

**Comment by stonecold123** [comment]
@kkusanagi 
Yes, the code snippet is inside an async function, and I don't use hot reloading. I have now moved on to use react-native-sqlite-storage to store the json objects. I can't get this specific method `AsyncStorage.getItem()` to work.

**Comment by fikrikarim** [comment]
Okay I got into this problem when testing with android device, but so far there isn't any issue with the iOS simulator.

Is this problem only exist in the development mode or is it also happening on production apps?

I guess this only happens in development right? Given that some of the comments before telling about the JS recompiling stuff?

**Comment by thevikas** [comment]
Having same problem while using expo client for my app on android. What happens is sqlite is not on the phone? Do I have to install sqlite with my app just so asyncstorage has a available storage?

**Comment by Khsed4** [comment]
Have the same problem on 55 version , and if someone knows an alternative solution , let us know @react-native-bot

**Comment by Poukai** [comment]
Will this ever be fixed ?

**Comment by Jalalx** [comment]
Sounds like no one from maintainers interested in this issue.

**Comment by Duckers** [comment]
Same problem here. Basic usage of `AsyncStorage` doesn't behave as expected. 

(Development setup, using a real iPhone X with live reload enabled)

**Comment by banli17** [comment]
same problem here，after await AysncStorage.getItem() 's code never executes

**Comment by IYTEC** [comment]
For me the problem started when I changed data in Asyncstorage, so I downloaded the Asyncstorage data, to know what data to remove, use Asyncstorage.removeItem to remove the specific data and restarted the application. 

Maybe Asyncstorage.clear don't actually clear. Hope it helps.

**Comment by Khsed4** [comment]
Hey Guys , I think I found the problem for myself ,it may seems stupid for you but what I did to work is just disable debug JS remotely and without logging my data inside AsyncStorage ,seems it's working every time that I need the data , so the solution for me was to disable Debug JS Remotely on Chrome , then it works perfectly , hopefully it solve other problem .

**Comment by talksik** [comment]
@Khsed4 Will try this but your solutions seems reasonable as maybe the cache is going to chrome? I know an instance of socket.io treating both the chrome debugger and my emulator as two different clients, and this may be a similar problem.

**Comment by RuedigerMoeller** [comment]
same here, debug mode emulator NO hot reload. seems timing dependent as on some machines it fails sometimes, on another machine it happens always

**Comment by guitar9** [comment]
+1

**Comment by Himujjal** [comment]
Same issue here for react native 0.55.4. 

What are some good AsyncStorage alternatives?

I will use the alternative until the fix is done. I used Realm but it is too complicated for simple usage.

**Comment by talksik** [comment]
I would say that just run `react-native run-android` again when you see issues pop up and they will be resolved. Also make sure that you don't use implementations from other people when you do a google search for how to store and receive. The newest page on Asyncstorage in React Native docs shows how you should have those async functions (i.e. how to properly do it).

**Comment by khuongdv** [comment]
Same issue here. Either getItem async or sync ways it doesn't work as expected. Many time it hangs but no error thrown.

**Comment by pstanton** [comment]
I get a "red screen" error: "the bind value at index 1 is null" - but nothing matching it in the console.

**Comment by radjivF** [comment]
Same problem here lol

**Comment by farzer** [comment]
do i have another way to replace asyncStorage for this issue?

**Comment by goughjo02** [comment]
I had basic asyncstorage commands (getItem, setItem) working last night. Then today they stopped working, which is how I got here. I am now using redux persist and this is successfully persisting the state. 

 I have experienced the not-working of asyncstorage - but for some reason redux-persist is working as expected.

    "react": "16.3.1",
    "react-native": "~0.55.2",
    "react-navigation": "^2.11.2",
    "react-redux": "^5.0.7",
    "redux": "^4.0.0",
    "redux-persist": "^5.10.0",

**Comment by Poukai** [comment]
So the solution here is to 

```
localdatabase.js
import { AsyncStorage } from 'react-native';
import _ from 'lodash';
export  const setItem= async(key, value) =>{
    try {
        return await AsyncStorage.setItem(key, JSON.stringify(value));
    } catch (error) {
        // console.error('AsyncStorage#setItem error: ' + error.message);
    }
}
export const getItem = async(key)=> {
    return await AsyncStorage.getItem(key)
        .then((result) => {
            if (result) {
                try {
                    result = JSON.parse(result);
                } catch (e) {
                    // console.error('AsyncStorage#getItem error deserializing JSON for key: ' + key, e.message);
                }
            }
            return result;
        });
}
export const  removeItem = async(key)=> {
    return await AsyncStorage.removeItem(key);
}
```
Use that file as : 

```
import {setItem,getItem} from './localdatabase';
getValuesFromDb= async()=>{
    const key1 = await getItem('1').then((result) => this.setState({ index1: result }));
    const key2 = await getItem('2').then((result) => this.setState({ index2: result }));
    }
  }
```

Make sure to run 

> yarn run android or ios everytime you make a change; This makes it easy to write the code and the async storage works without any issues. Also when building for production it works like charm the issue is when you connect the debugger or enable hot loading it fucks up.

**Comment by RobTS** [comment]
Same here, and literally forever. So I need to close the app on Android and open it again every time I reload -.-

**Comment by bhanuc** [comment]
Me and my team of 3 RN developers had simultaneously gotten stuck in the same issue, so I wrote a simple FIleSystem based alternative to AsyncStorage named [react-native-fs-store](https://www.npmjs.com/package/react-native-fs-store) module, which is drop-in replacement of asyncStorage for very simple use cases. It also allows automatic switch to actual AsyncStorage in production. I hope it unblocks someone else as well.

**Comment by rkulinski** [comment]
```

**Comment by crimson-med** [comment]
Same problem here. 
I do a simple `setItem` then try a `getItem`

It just returns:

```json
{
"_40":0,
"_65":0,
"_55":null,
"_72":null,
}
```
But as mentioned by many people for the maintainers it seems storage isn't important and won't be fixed anytime soon. 

I recommend using:

[MongoDB](https://www.npmjs.com/package/react-native-local-mongodb)
[SQLite](https://www.npmjs.com/package/react-native-sqlite-storage?activeTab=readme)

**Comment by goughjo02** [comment]
That's a Promise response. Try `getItem(ITEMKEY).then(res => console.log(res))`.......... or did you try that?

**Comment by crimson-med** [comment]
@goughjo02 same as well I tried but sadly my value is never returned hence I have switched to mongo and it works perfect now

**Comment by 4E71-NOP** [comment]
Hmmm i had the same issue. 

I tried many different syntaxes you probably got in the previous posts.  No success so far. At first i was incline to throw it away.  Then i tried less complex (imbrication-wise) syntaxes. 

I got there and it finally worked for me.

```js
    async componentDidMount() {
        var AsyncVal = await AsyncStorage.getItem("lchMoteurRch");
        await this.setState({ lchMoteurRch: AsyncVal });
        console.log("ControllerXXX/componentDidMount : state.lchMoteurRch=" + this.state.lchMoteurRch + "; AsyncVal=" + AsyncVal );
    }
// -> ControllerXXX/componentDidMount : state.lchMoteurRch=abcdef; AsyncVal=abcdef

```
and

```js
    async myfunc(itemValue, itemIndex) { // <- from a Picker
        await AsyncStorage.setItem('lchMoteurRch', itemValue);
        await this.setState({ lchMoteurRch: itemValue }); //shown here if you need to set some state.xxx
        console.log("ControllerXXX/myfunc : index=" + itemIndex + "; val=" + itemValue + "; state.lchMoteurRch=" + this.state.lchMoteurRch)
    }
```
Version:

- react-native-cli: 2.0.1
- react-native: 0.56.0

with gradle configured like this:
- buildToolsVersion = "27.0.3"
- minSdkVersion = 16
- compileSdkVersion = 27
- targetSdkVersion = 27
- supportLibVersion = "27.0.1"


It works in **"normal debug"** mode and **variant=release** for me.

Apparently minimizing the callbacks usage probably helped. It may not be that elegant but it's working properly. That's not a bad bargain all things considered.

Now, it raises questions about "this.setState" callback. While debugging i was using the setState callback parameter to log to the console the 'state' object supposedly updated values but with no success. 

The asyncStorage value was saved, retrieved and passed properly. I could log it to the console in the 'then' statement. 
something like : AsyncStorage.setItem('lchMoteurRch', itemValue).then(... ...);

The setState was executing (in a '.then()' statement) but for some reason it did not update the state object. I know the setState method is particular amongst the asynchronous methods. But i wasn't expecting that. Maybe:
- 1 The setState fired the callback too soon which is weird as it IS supposed to be a callback.
- 2 The setState did not update anything because it did not execute. Or it was not in the right place / context. Maybe a scope issue... dunno.

I'm left only with question on that. Is it an isolated issue or should we expect different callback behaviors depending on the function used ? Then, where's the doc ???


Well... The solution i found is supposed to do exactly the same. Still... this one worked for me.

**Comment by sinoridha** [comment]
I face this issue while I run on this condition : 
- Remote JS debugging enable 
- run on an emulator with: `npm run android`
- "react-native": "~0.55.2"

this solution works for me:
1. Stop remote JS debugging.
2. close the app on the emulator.
3. run with `react-native run-android`
4. enable remote JS debugging to check the issue.

**Comment by mezod** [comment]
Experiencing this issue in production with a single AsyncStorage.getItem. Sometimes it works, sometimes it just returns null.

**Comment by fikrikarim** [comment]
Can we update status of issue? 0.57 is released now, and from the [changelog](https://github.com/react-native-community/react-native-releases/blob/master/CHANGELOG.md#android-specific-fixes) they mentioned AsyncStorage fix. Is the change solve this issue?

**Comment by Khsed4** [comment]
Did anyone test 0.57 version ? if it's solved there or no ? they promised that it would be solved.

**Comment by apcjs** [comment]
@Khsed4, yes I have tested my app on 0.57 and it works fine. I have had the problem with previous versions consistently.

**Comment by fikrikarim** [comment]
@apcjs, nice. I think we can close this issue now?

**Comment by Ovski4** [comment]
I'm trying to make this work for a very basic app that I test on expo.
I'm very new to this so pardon me if I'm wrong but for what I understand, It won't work with expo as:
- since version 0.56 React Native is using Babel 7 and babel-preset-expo doesn't support Babel 7 (found here: https://github.com/expo/expo/issues/1954).
- the fact that version 0.57 is not present in this file: https://github.com/react-community/create-react-native-app/blob/master/VERSIONS.md does not help.
This might be fine for those who use an emulator or an android device directly

**Comment by kalpeshjjethva** [comment]
you can check this.  [storage](https://stackoverflow.com/a/52331514/4845562)

**Comment by sagenate24** [comment]
I had this problem until I removed all console logs in both my asynchronous functions and logger middleware!

**Comment by farzer** [comment]
but in 0.57, can you build your android apk? @apcjs @fikrikarim

**Comment by apcjs** [comment]
@farzer yes, I'm able to build an APK using 0.57. This was never a problem for me in previous versions. The issue I had was with reading a value from asyncStorage once you reload the app using RR. The promise never got fulfilled. This was consistently failing for me. With 0.57, it works fine. 

Also note that I did not upgrade from 0.55.4 to 0.57. I ran react-native init and copied my source code from the old folder. Not sure if that makes a difference.

**Comment by farzer** [comment]
for me, upgrade to 0.57 meet another problem. see #19211

**Comment by Rana-Muhammad-Amir** [comment]
by converting data to stringify and then setItem ......
and after getting it converting it back work for me ....
setting data:
 AsyncStorage.setItem('saveData', JSON.stringify(value));
getting data:
 AsyncStorage.getItem('saveData').then((value) => {
   
        data=JSON.parse(value);
}).done();

worked for me...

**Comment by sunithakk** [comment]
> Hey Guys , I think I found the problem for myself ,it may seems stupid for you but what I did to work is just disable debug JS remotely and without logging my data inside AsyncStorage ,seems it's working every time that I need the data , so the solution for me was to disable Debug JS Remotely on Chrome , then it works perfectly , hopefully it solves other problem .

after build APK again facing same issue. is there any other alternative solution?

**Comment by Erwin0Maleki** [comment]
0.57.1 not fixed yet...

**Comment by Khsed4** [comment]
@sunithakk sorry for my late reply.
I was stuck with this stupid problem for almost one week , no way that is suggested in all communities worked for me , but then I found something that is built in with **setState()** function , which it has a callback function as this website says : https://medium.learnreact.com/setstate-takes-a-callback-1f71ad5d2296.
so the only way that I guarantee that it's the only secure way so far is this that u use the **setState()** function in your promise and everything that you need to run , put them on a function and call it for the **setSate() callback function** , this is only way you can make sure yourself that you neither get null nor never calling the function . Here I'm going to provide an example for it .
` try {
      AsyncStorage.getItem('firebase_token', (err, item) => {
        if (item) {
          this.setState({
            firebase_token: item,
          }),this.tokenToServer();
        }
      });
     

    } catch (error) {
      console.log("Error retrieving data" + error);
    }`

**Comment by wesleyguirra** [comment]
same here

**Comment by fuj1w4r4** [comment]
I have upgraded react-native to 0.57. The problem still persist, happen sporadically. and some devices not working.

**Comment by sandersjj** [comment]
Hello everyone,

Is there a workaround for this issue?

Thanks.

**Comment by sandersjj** [comment]
Hello,

Is there a workaround for this issue?

Thanks.

**Comment by pshah123** [comment]
@sandersjj try Expo SecureStorage. Alternatively, make sure you eject from expo, and then set and get a very simple standard key when your app loads. For some reason if you can get a full cycle in, it works perfectly (I thin k it has something to do with initializations and failures being blocking on the UI thread, but have no real clue why this works).

**Comment by wesleyguirra** [comment]
@pshah123 how to use Expo SecureStorage in a native app?

**Comment by rustan-id** [comment]
Still face the same issue

  async componentDidMount() {
    try {
      let cities = await AsyncStorage.getItem(key);
      cities = JSON.parse(cities);
      this.setState({ cities });
    } catch (e) {
      console.log("error from AsyncStorage: ", e);
    }
  }

$ react-native --version
react-native-cli: 2.0.1
react-native: 0.57.4

**Comment by sunithakk** [comment]
> > @sunithakk sorry for my late reply.
> > I was stuck with this stupid problem for almost one week , no way that is suggested in all communities worked for me , but then I found something that is built in with **setState()** function , which it has a callback function as this website says : https://medium.learnreact.com/setstate-takes-a-callback-1f71ad5d2296.
> > so the only way that I guarantee that it's the only secure way so far is this that u use the **setState()** function in your promise and everything that you need to run , put them on a function and call it for the **setSate() callback function** , this is only way you can make sure yourself that you neither get null nor never calling the function . Here I'm going to provide an example for it .
> > ```
> > ` try {
> >       AsyncStorage.getItem('firebase_token', (err, item) => {
> >         if (item) {
> >           this.setState({
> >             firebase_token: item,
> >           }),this.tokenToServer();
> >         }
> >       });
> >       } catch (error) {
> >       console.log("Error retrieving data" + error);
> >     }`
> > ```
> 
> Actually m storing data in separate javascript file. not inside component. pls find below code.
> import React from 'react';
> import { AsyncStorage, Text, View, TextInput} from 'react-native'
> var SQLite = require('react-native-sqlite-storage')
> 
> const LocalStorage = {
> retrieveData : async (value) => {
> try {
> const value = await AsyncStorage.getItem('myKeySpl');
> 
> ```
>   if (value !== null) {
> 
>     return JSON.parse(value);
>   }else{
>     return false;
>   }
>  } catch (error) {
>   return false; 
>  }
> ```
> }
> 
> }
> 
> module.exports = LocalStorage;
> your solution is not working here....

> `kkkkkk`

>

**Comment by NativLouie** [comment]
i am also experiencing this problem any updates?

**Comment by pascaloliv** [comment]
Hello everyone, same issue here with RN 56.0.

```
[...]
console.warn('Before user_id')
const user_id = await AsyncStorage.getItem(USER_ID)
console.warn('After user_id')
[...]
```

App log "Before user_id" and can't go further. It looks like the AsyncStorage promise is never resolved in some case. Which is really blocking. Anyone using an alternative to AsyncStorage ?

**Comment by rustan-id** [comment]
i think realm is the best choice.

2018-11-29 23:16 GMT+07.00, Olivier PASCAL <notifications@github.com>:
> Hello everyone, same issue here with RN 56.0.
>
> ```
> [...]
> console.warn('Before user_id')
> const user_id = await AsyncStorage.getItem(USER_ID)
> console.warn('After user_id')
> [...]
> ```
>
> App log "Before user_id" and can't go further. It looks like the
> AsyncStorage promise is never resolved in some case. Which is really
> blocking. Anyone using an alternative to AsyncStorage ?
>
> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/facebook/react-native/issues/18372#issuecomment-442892950


-- 
Rustan, S.Si.

**Comment by danfsd** [comment]
To all of those still experiencing the issue, the `SecureStore` from **Expo** is a viable alternative, which has APIs signatures very similar to `AsyncStorage` and already comes bundled with **Expo**. That solved the issue for me.

**Comment by ditorahard** [comment]
I'm also facing the same issue. Is there any update on this or maybe a workaround (other than using expo version) ? Thanks

**Comment by Sukumar-Abhijeet** [comment]
I also encountered the same issue. I was binding an async function in the UI of  one screen. So whenever I was navigating to that screen, the app was getting stuck and restarting the app was not working. It was working only if i restart the expo app and then the project app.

Please check whether an async function is getting invoked on render or not. That could be the issue of AsyncStorage not fetching any item because the previous async has not completed yet. Thanks.

**Comment by skout90** [comment]
same issue......
in rn 0.57.8

**Comment by jclwong** [comment]
This resolved the issue for me. On a related thread: https://github.com/facebook/react-native/issues/14101#issuecomment-345563563

> After further investigation: in my case I had an AsyncTask in some Java code (a local web server in my case) that was blocking and hence blocking other AsyncTasks (which AsyncStorage relies on). Setting my AsyncTask to execute in parallel to other tasks (via myTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR) instead of myTask.execute()) fixed the issue.

My native Android code had reference to an AsyncTask, which for some reason mucks with AsyncStorage. Once removed it works as intended.

**Comment by Bradzer** [comment]
Same issue for me. 
It is such a shame that a such critical and old issue is not being solved.

**Comment by robertpitt** [comment]
Experiencing the same issue, AsyncStorage is just unusable.

**Comment by Bradzer** [comment]
@hramos why is the bug report label removed when the issue is not resolved :(
It's a really critical issue that needs more attention than it has been given so far

**Comment by yashojha19** [comment]
Hellp all,
I have had this issue once too making the app in production go freeze. what
a day that was.
what worked out for me for temporary solution was instead of getting and
setting value from asyncstorage. I used React natie file system (RNFS) and
used it as reading and writing file in app reserved root space which is not
accessible through file managers, and you're anyway going to store auth
tokens here or even you can encrypt/decrypt your stringified json there.

**Comment by Georgge** [comment]
Problem is not `getItem`, the problem is `setItem` or that I think.
The difference between `getItem` and `setItem` is that set not return something (except a null). So, the `await` of `setItem` never pass to the next set. Actually, even never set the data.

So, the solution (at least for me) is: **not to use `await` in` setItem`.**

**Comment by eightyfive** [comment]
At last some valuable information! Thanks @Georgge

**Comment by AleksandarSavic95** [comment]
iOS simulator (iPhone X) works perfectly, but on a real iPhone 6, it fails (storage is empty after going to another screen or reloading).

**Comment by cpojer** [comment]
This issue has been moved to react-native-community/react-native-async-storage#12.

**Comment by chenop** [comment]
To fix the live reload - switching to async await worked for me.

Checked this [SO Answer](https://stackoverflow.com/a/54970518/1846993)

**Comment by murilokaos** [comment]
Have you remembered to put the item in asyncstorage as string (JSON.stringify) and when reading the parse item (JSON.parse) ?? this may be the mistake of many.

**Comment by jeffelector** [comment]
> Problem is not `getItem`, the problem is `setItem` or that I think.
> The difference between `getItem` and `setItem` is that set not return something (except a null). So, the `await` of `setItem` never pass to the next set. Actually, even never set the data.
> 
> So, the solution (at least for me) is: **not to use `await` in` setItem`.**

great !!! you save my day

**Comment by jorodriguez** [comment]
//Login.js
..
onClickListener = (viewId) => { 
       AsyncStorage.setItem('logeado', JSON.stringify(this.usuario.auth));
       AsyncStorage.setItem('usuario', JSON.stringify(this.usuario));
}

..

//Welcome.js

...
constructor(props) {
    super(props);

    this.state = {
      loading: false,
       usuarioSesion : null
    }
    this._bootstrapAsync();
  }
     
  _bootstrapAsync = async () => {    
      const usuario = await AsyncStorage.getItem('usuario');
       Alert.alert("json",JSON.parse(usuario));
      this.setState({usuarioSesion :usuario});
  };

  salir = () => { 
    AsyncStorage.removeItem('logeado');
    AsyncStorage.removeItem('usuario');
    this.props.navigation.navigate('AuthLoading', {});
  }


its works for me.

**Comment by paddy57** [comment]
I don't know what will be solution in your case but this one worked for me 

Firstely i was importing some componets from react-native like this 
import { Text, StyleSheet, TouchableHighlight, TextInput, } from 'react-native';
import {  AsyncStorage } from 'react-native';

and the AsyncStorage.getItem or AsyncStorage.setItem was not working, whatever code was there after AsunStorage was not running, so

I Modified my import like to this 
import { Text, StyleSheet, TouchableHighlight, TextInput, AsyncStorage  } from 'react-native';

and now everything was working fine.
I don't know how it got resolved, but my  AsyncStorage.setItem and AsyncStorage.getItem both was working fine

AsyncStorage.setItem('DEMO', 'false');

  try {
            let data = await AsyncStorage.getItem("DEMO");
            alert(data);
        }
        catch (error) {

        }

**Comment by kamirt** [comment]
Hi everyone, 

> Problem is not `getItem`, the problem is `setItem` or that I think.
> The difference between `getItem` and `setItem` is that set not return something (except a null). So, the `await` of `setItem` never pass to the next set. Actually, even never set the data.
> 
> So, the solution (at least for me) is: **not to use `await` in` setItem`.**

Not just this I think. I've used your solution, but `getItem('key', callback)` and `getItem('key').then()` also doesn't work. Also it does'nt work directly inside `componentDidMount` hook. It become working when I've took this away to another function like this:
`_getItem = async function () {
  const result = await AsyncStorage.getItem('key')
  return result
}
componentDidMount () {
  this._getItem().then((result) => {
    <so something with result>
})
}`

**Comment by b3hz4d** [comment]
I found solution for this
Use this link https://github.com/react-native-community/async-storage/blob/LEGACY/docs/advanced/DedicatedExecutor.md

**Comment by irishabhyadav90** [comment]
Even the custom Hooks (UseStorage) in React Native causing the same Problem.
 

import {useState, useEffect} from 'react';
import {AsyncStorage} from '@react-native-community/async-storage';

/* Hook for AsyncStorage */

const useStorage = (key, defaultValue) => {
  const [storage, updateStorageValue] = useState(defaultValue, '');
  async function getStorageValue() {
    let value = key || defaultValue;

    try {
      value = JSON.parse(await AsyncStorage.getItem(key)) || defaultValue;
    } catch (e) {
    } finally {
      updateStorageValue(value);
    }
  }

  async function updateStorage(newValue) {
    try {
      if (newValue === null) {
        await AsyncStorage.removeItem(key);
      } else {
        const value = JSON.stringify(newValue);
        await AsyncStorage.setItem(key, value);
      }
    } catch (e) {
    } finally {
      updateStorageValue(newValue);
    }
  }

  useEffect(() => {
    getStorageValue();
  }, [getStorageValue]);

  return [storage, updateStorage];
};

export default useStorage;
